File: COURSE 1 (Algorithmic Toolbox)/week2_algorithmic_warmup/week 2 assignment/src/com/company/FibonacciHuge.java
package com.company;

import java.util.*;

public class FibonacciHuge {
    private static int Pisano(long n){
        long previous = 0;
        long current  = 1;

        for (long i = 0; i < n*n; ++i) {
            long tmp_previous = previous;
            previous = current;
            current = (tmp_previous + current)%n;
            if(previous==0 && current==1){
                return (int) (i+1);
            }
        }
        return -1;
    }

    private static long getFibonacciHuge(long n, long m) {
        n=n%Pisano(m);

        if(n<=1){
            return n;
        }

        long previous = 0;
        long current  = 1;

        for (long i = 0; i < n-1; ++i) {
            long tmp_previous = previous;
            previous = current;
            current = (tmp_previous + current)%m;
        }
        return current%m;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        long n = scanner.nextLong();
        long m = scanner.nextLong();
        System.out.println(getFibonacciHuge(n, m));
    }
}

File: COURSE 1 (Algorithmic Toolbox)/week2_algorithmic_warmup/week 2 assignment/submitted files c++/fibonacci_huge.cpp
#include <iostream>

long long Pisano(long long n){
        long long previous = 0;
        long long current  = 1;

        for (long long i = 0; i < n*n; ++i) {
            long long tmp_previous = previous;
            previous = current;
            current = (tmp_previous + current)%n;
            if(previous==0 && current==1){
                return (i+1);
            }
        }
        return -1;
    }

long long get_fibonacci_huge_naive(long long n, long long m) {
    n=n%Pisano(m);

    if (n <= 1)
        return n;

    long long previous = 0;
    long long current  = 1;

    for (long long i = 0; i < n - 1; ++i) {
        long long tmp_previous = previous;
        previous = current;
        current = (tmp_previous + current)%m;
    }

    return current % m;
}

int main() {
    long long n, m;
    std::cin >> n >> m;
    std::cout << get_fibonacci_huge_naive(n, m) << '\n';
}
